### ISSUE TITLE
Extending salt-cloud profile/provider config using a list format

### ISSUE THREAD

**Issue by jrsdav** [saltstack/salt#40897]
### Description of Issue/Question
Using a list format with the extend feature in a cloud profile/provider configuration file results in the following error:

```
salt-cloud -p app1-a.e TEST-app1
/usr/lib/python2.7/site-packages/salt/config/__init__.py:2617: DeprecationWarning: The term 'provider' is
being deprecated in favor of 'driver'. Support for 'provider' will be removed in Salt Nitrogen. Please con
vert your cloud provider configuration files to use 'driver'.
[ERROR   ] unhashable type: 'list'
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 200, in parse_args
    process_option_func()
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 516, in process_config_dir
    self.config.update(self.setup_config())
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 3022, in setup_config
    return config.cloud_config(self.get_config_file_path())
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2231, in cloud_config
    providers_config)
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2398, in vm_profiles_config
    return apply_vm_profiles_config(providers, overrides, defaults)
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2463, in apply_vm_profiles_config
    if extends not in vms:
TypeError: unhashable type: 'list'
Usage: salt-cloud [options] <-m MAP | -p PROFILE> <NAME> [NAME2 ...]

salt-cloud: error: Error while processing <bound method SaltCloud.process_config_dir of <salt.cloud.cli.Sa
ltCloud object at 0x2c981d0>>: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 200, in parse_args
    process_option_func()
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 516, in process_config_dir
    self.config.update(self.setup_config())
  File "/usr/lib/python2.7/site-packages/salt/utils/parsers.py", line 3022, in setup_config
    return config.cloud_config(self.get_config_file_path())
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2231, in cloud_config
    providers_config)
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2398, in vm_profiles_config
    return apply_vm_profiles_config(providers, overrides, defaults)
  File "/usr/lib/python2.7/site-packages/salt/config/__init__.py", line 2463, in apply_vm_profiles_config
    if extends not in vms:
TypeError: unhashable type: 'list'
```

### Setup
Here's an example of my cloud profile config:

```
# App Top Level Profiles
 app1-a.e:
  extends:
   - a.external
   - app1
   - ubuntu

# App Base Level Profiles
 app1:
  size: m4.xlarge
  network_interfaces:
     SecurityGroupId:
        - sg-123456
        - sg-7891011
  minion:
    grains:
      roles:
        - app1
  tag:
    Roles: app1

 app2:
  size: m4.xlarge
  network_interfaces:
     SecurityGroupId:
        - sg-123456
        - sg-7891011
  minion:
    grains:
      roles:
        - app2
  tag:
    Roles: app2
 
# Subnet Level Profiles
 a.external:
  provider: aws-provider
  network_interfaces:
      SubnetId: subnet-1234567

 b.external:
  provider: aws-provider
  network_interfaces:
      SubnetId: subnet-1234567

 c.external:
  provider: aws-provider
  network_interfaces:
      SubnetId: subnet-1234567

# Base Level Profiles    #
 ubuntu:
  image: ami-1234567
  ssh_interface: private_ips
  ssh_username: ubuntu
  private_key: /etc/salt/private_key.pem
  script_args: stable 2016.11.1
  ebs_optimized: True
  network_interfaces:
    - DeviceIndex: 0

 centos:
  image: ami-1011121
  ssh_interface: private_ips
  ssh_username: centos
  private_key: /etc/salt/private_key.pem
  script_args: stable 2016.11.1
  ebs_optimized: True
  network_interfaces:
    - DeviceIndex: 0
```

### Why This Would Be Useful
By extending your configuration in a list format, you could make your cloud profiles more modular.  This would be helpful when you find yourself having to repeat persistent settings for multiple options, such as AWS availability zones, VPC subnets, and AMI images.

### Versions Report
```
sudo salt --versions-report
Salt Version:
           Salt: 2016.11.1

Dependency Versions:
           cffi: 1.9.1
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.21.4
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
         pygit2: 0.21.4
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.3.1611 Core
        machine: x86_64
        release: 3.10.0-514.6.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core
```

**Comment by wes-novack** [comment]
+1

**Comment by Ch3LL** [comment]
So support for more then one extends? I think thats a good idea. I'll approve as a feature request.

**Comment by jrsdav** [comment]
Yep! Thanks Megan!

Just to note for anyone referencing this in the future, according to the documentation you can do a nested extend, but with limitations:

>Note
> Extending cloud profiles and providers is not recursive. For example, a profile that is extended by a second profile is possible, but the second profile cannot be extended by a third profile.
https://docs.saltstack.com/en/latest/topics/cloud/config.html#extending-profiles-and-cloud-providers-configuration

This would be different in that you're extending multiple profiles within one profile, but doing so in a list.

**Comment by cetanu** [comment]
Very interesting... How do you propose order and priority be handled? I'm guessing that the first profile in the list has its keys overwritten by subsequent items?

**Comment by jrsdav** [comment]
Yep! That's the behavior I've seen during my tests (using 2016.11.1). The keys from the last profile matched against are the one used for the salt-cloud bootstrap process.

**Comment by cetanu** [comment]
Have you tried this on the develop version? I'm looking through the code and it seems doable... even nested extends, which is a little bit mind blowing

**Comment by jrsdav** [comment]
I have not yet had the time to test in the dev version, unfortunately. But thanks for the heads up! I'd love for you to report back if you get around to testing before I do.

**Comment by cetanu** [comment]
Turns out I was wrong, I tried to get this working yesterday and today but I can't wrap my head around the function: apply_vm_profile_config

In my own projects I would probably be re-writing this function, as it appears to be a bit of a mess, and it's exact purpose is somewhat unclear.

I'm on Windows right now so I'll have to create a linux machine to run the unit tests and see if I can just do a rewrite. But it's a bit extreme, and what if the unit tests don't cover all the cases? I'm worried.
